app/controllers/admin/admin_controller.rb

```diff
@@ -0,0 +1,64 @@
+module Admin
+  class AdminController < ApplicationController
+    before_action :authorized?
+
+    def index
+      @source_models = source_model.all
+    end
+
+    def new
+      @source_model = source_model.new
+    end
+
+    def create
+      @source_model = source_model.new(send("#{generate_source_str_s}_params"))
+
+      if @source_model.save
+        redirect_to send("admin_#{generate_source_str_p}_path")
+      else
+        render 'new'
+      end
+    end
+
+    def show
+    end
+
+    def edit
+    end
+
+    def update
+      if @source_model.update(coupon_params)
+        redirect_to send("admin_#{generate_source_str_p}_path")
+      else
+        render 'edit'
+      end
+    end
+
+    def destroy
+      @source_model.destroy
+
+      redirect_to send("admin_#{generate_source_str_p}_path")
+    end
+
+    private
+
+    def authorized?
+      if current_user.nil? or !current_user.admin?
+        flash[:error] = "You are not authorized to view that page."
+        redirect_to '/'
+      end
+    end
+
+    def generate_source_str_s
+      controller_name.singularize
+    end
+
+    def generate_source_str_p
+      controller_name
+    end
+
+    def source_model
```

Comment:
also you should do `@source_klass ||= controller_name.singularize.classify.constantize`

which is just a shortened version of 

```
if @source_klass
  return @source_klass
else 
  @source_klass = controller_name.singularize.classify.constantize
  return @source_klass
end
```

and it prevents calling the logic over and over as you call source_model multiple times during execution